### PR TITLE
Port Variational Inference with Adversarial Learning (VITS) to NeMo TTS

### PR DESCRIPTION
**Pull request by jasonjjl1999**
# Overview

This change adds [Variational Inference with adversarial learning (VITS)](https://github.com/jaywalnut310/vits) to the collection of TTS models.

VITS uses the `monotonic_align` module, which requires Cython to be installed prior to building (i.e. run `pip install cython` before running `./reinstall.sh`).

# Tasks

- [x] Fix errors in audio and spectrogram dimensions
- [ ] Fix logging for validation loss
- [ ] Verify training performance
- [ ] Verify quality of TTS output
- [ ] Refactor code and cleanup to be more consistent with other models

### CODE REVIEW

**Comment by jasonjjl1999** [comment]
# Tasks

- [ ] Fix errors in audio and spectrogram dimensions
- [ ] Fix logging for validation loss
- [ ] Verify training performance
- [ ] Verify quality of TTS output
- [ ] Refactor code and cleanup to be more consistent with other models

File: examples/tts/conf/vits.yaml [hunk 1]
```diff
@@ -0,0 +1,256 @@
+name: "VITS"
+
+labels: [' ', '!', "'", '(', ')', ',', '-', '.', ':', ';', '?', 'a', 'b', 'c', 'd', 'e', 'f', 'g',
+         'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z']
```
Comment (blisc) [inline examples/tts/conf/vits.yaml:4]: There are some parameters you can clean up like `labels` and `self._parser` in the model since they are unused.

File: examples/tts/conf/vits.yaml [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+  num_nodes: 1
+  accelerator: ddp
+  accumulate_grad_batches: 1
+  checkpoint_callback: False  # Provided by exp_manager
+  logger: False  # Provided by exp_manager
+  # gradient_clip_val: 1000.0
+  flush_logs_every_n_steps: 1000
+  log_every_n_steps: 100
+  check_val_every_n_epoch: 5
+  precision: 16
```
Comment (blisc) [inline examples/tts/conf/vits.yaml:250]: FYI, you should use `precision: 16`, only if you are using a Volta or Ampere GPU. Otherwise, it's better to remove this line

File: nemo/collections/tts/models/vits.py [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+            # _ = cfg.model.train_ds.manifest_filepath
+            _ = cfg['dataset']['manifest_filepath']
+        except omegaconf.errors.MissingMandatoryValue:
+            logging.warning("manifest_filepath was skipped. No dataset for this model.")
+            return None
+
+        dataset = instantiate(cfg.dataset)
+        return torch.utils.data.DataLoader(  # noqa
+            dataset=dataset, collate_fn=dataset.collate_fn, **cfg.dataloader_params,
+        )
```
Comment (blisc) [inline nemo/collections/tts/models/vits.py:404]: What does this portion of the code do?

File: nemo/collections/tts/models/vits.py [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+        y_hat, attn, mask, *_ = self.net_g.infer(x, x_lengths, max_len=1000)
+        y_hat = y_hat.squeeze()
+        y_hat_lengths = mask.sum([1, 2]).long() * self._cfg.train_ds.dataset.hop_length
+
+        # spec = self.get_spec(y)
+        # spec_lengths = torch.ones(spec.shape[0]) * spec.shape[2]
+
+        # Note to modify the functions / use the ones in NeMo, we need the lengths
+        mel, mel_lengths = self.audio_to_melspec_precessor(y, y_lengths)
+        y_hat_mel, y_hat_mel_lengths = self.audio_to_melspec_precessor(y_hat, y_hat_lengths)
```
Comment (blisc) [inline nemo/collections/tts/models/vits.py:285]: What is the difference between `self.audio_to_melspec_precessor`, and the `modules.spec_to_mel_torch`/`modules.mel_spectrogram_torch` functions?

File: nemo/collections/tts/models/vits.py [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+                spec = torch.view_as_real(spec)
+            spec = torch.sqrt(spec.pow(2).sum(-1) + 1e-9)
+        return spec
+
+    def training_step(self, batch, batch_idx):
+        # (x, x_lengths, y, y_lengths) = batch
+        (y, y_lengths, x, x_lengths) = batch
+
+        spec = self.get_spec(y)
+        spec_lengths = torch.ones(spec.shape[0]) * spec.shape[2]
```
Comment (blisc) [inline nemo/collections/tts/models/vits.py:198]: As I mentioned in the meeting, you need to get accurate spec_lengths

Comment (renrichard) [reply nemo/collections/tts/models/vits.py:198]: done.

Comment (renrichard) [reply nemo/collections/tts/models/vits.py:198]: done.

File: examples/tts/conf/vits.yaml (hunk 2 above)
Comment (jasonjjl1999) [reply examples/tts/conf/vits.yaml]: Removed in df7e99660e5c7f957fdbec153dbeb1b239579ce6

**Comment by blisc** [comment]
close in favour of #3595
